File: apps/web/src/api/todos.ts
/**
 * 待办清单的接口封装。
 *
 * ## PATCH 的类型为什么写成 `?: string | null`
 *
 * 后端靠"字段**有没有出现**在请求体里"来区分两件事：
 *
 * | 前端给的 | JSON 里 | 后端行为 |
 * |---|---|---|
 * | 键**不出现** | `{}` | 这一项不动 |
 * | `due_date: null` | `{"due_date":null}` | **清空**截止日 |
 * | `due_date: undefined` | `{}`（被 `JSON.stringify` 丢掉） | 这一项不动 |
 *
 * 所以「清空」必须用 `null`，**不能用 `undefined`** ——
 * `JSON.stringify` 会把 `undefined` 的键整个删掉，于是"清空"静默变成"不动"。
 * 类型里保留 `| null` 就是为了让"传 null"这件事在调用方显式可见。
 */

import { deleteJson, getJson, patchJson, postJson } from './http'

export interface TodoItem {
  id: number
  title: string
  completed: boolean
  /** `YYYY-MM-DD`；null = 没有截止日 */
  due_date: string | null
  /** `none` 不计时 / `countup` 正计时 / `countdown` 倒计时 */
  timer_mode: TimerMode
  /** 只有倒计时才有值 */
  timer_minutes: number | null
  /** **累计**已计时秒数（不是单次时长） */
  spent_seconds: number
  created_at: string
  updated_at: string
}

/** 计时方式。与后端 `app/models/todo.py` 的 `TimerMode` 一一对应。 */
export type TimerMode = 'none' | 'countup' | 'countdown'

export interface TodoList {
  items: TodoItem[]
  total: number
}

export interface TodoListQuery {
  /** 不传 = 全部；true = 只看已完成；false = 只看未完成 */
  completed?: boolean
  limit?: number
  offset?: number
}

/**
 * 新建。`due_date` 省略或给 null 都表示"没有截止日"。
 *
 * ⚠️ 计时那两个字段是**一对**：给了 `timer_mode` 就要一起给 `timer_minutes`
 * （倒计时给分钟数，其余给 `null`）。后端会校验自洽性，不自洽 → 422 ✓
 */
export function createTodo(
  title: string,
  options: { dueDate?: string | null; timerMode?: TimerMode; timerMinutes?: number | null } = {},
): Promise<TodoItem> {
  const mode = options.timerMode ?? 'none'
  return postJson<TodoItem>('/api/todos', {
    title,
    ...(options.dueDate ? { due_date: options.dueDate } : {}),
    ...(mode !== 'none' ? { timer_mode: mode, timer_minutes: options.timerMinutes ?? null } : {}),
  })
}

export function listTodos(query: TodoListQuery = {}): Promise<TodoList> {
  const params = new URLSearchParams()
  if (query.completed !== undefined) params.set('completed', String(query.completed))
  if (query.limit !== undefined) params.set('limit', String(query.limit))
  if (query.offset !== undefined) params.set('offset', String(query.offset))
  const suffix = params.toString()
  return getJson<TodoList>(`/api/todos${suffix ? `?${suffix}` : ''}`)
}

/**
 * 改一条。**只把要改的键放进 `changes`**：
 * 没放进来的键后端不会动（包括 `due_date` 与计时配置）。
 */
export interface TodoChanges {
  title?: string
  completed?: boolean
  /** 显式传 `null` 才是**清空**；不传这个键 = 不动 */
  due_date?: string | null
  /** ⚠️ 改计时必须**成对**给：`timer_mode` + `timer_minutes` 一起 */
  timer_mode?: TimerMode
  timer_minutes?: number | null
  /** 计时暂停 / 收尾时回写累计秒数 */
  spent_seconds?: number
}

export function updateTodo(id: number, changes: TodoChanges): Promise<TodoItem> {
  return patchJson<TodoItem>(`/api/todos/${id}`, changes)
}

export function deleteTodo(id: number): Promise<{ deleted: boolean; todo_id: number }> {
  return deleteJson<{ deleted: boolean; todo_id: number }>(`/api/todos/${id}`)
}

File: apps/api/app/api/routes/todos.py
"""待办事项接口。

## 四个端点，都在自己的资料之外

```
POST   /todos            新建
GET    /todos            列出（可按完成状态过滤）
PATCH  /todos/{todo_id}  改（标题 / 完成状态 / 截止日）
DELETE /todos/{todo_id}  删
```

**刻意不做 `GET /todos/{todo_id}`**：列表已经带回全部字段，
行内编辑用的是列表里那一条，单独取一条没有调用方。
少一个端点就是少一处越权面 —— 这一条在 5D 的方案里已经确认过。

## 为什么每个端点都写 `require_learner_id`

`current_learner_id` 在未登录时会**回退到 `DEFAULT_LEARNER_ID = "local"`**，
而演示账号 demo 的 `learner_id` **也是 `local`** —— 于是"未登录访客 = demo 账号"。
对"本机自己用"的场景这没问题；但待办是**私人内容**，
用 `current_learner_id` 就意味着任何人都能看到 demo 的待办。
所以这里与 `study.py` 的所有端点保持同一把门：**必须登录**。

## 归属校验

单条操作全部走 `todo_service.get_owned(db, todo_id, learner_id)`，
取不到（**不存在、或不属于自己**）统一抛 `TodoNotFound` → 这里转 **404**。
不返回 403 —— 403 等于确认"这个 id 存在"，那是一个枚举接口。
"""

from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from app.api.deps import require_learner_id
from app.db.session import get_db
from app.schemas.todo import (
    TodoCreate,
    TodoItem,
    TodoListResponse,
    TodoUpdate,
)
from app.services import todo_service

router = APIRouter(prefix="/todos", tags=["todos"])

#: "不存在或不属于你" 的统一说法。**两种情况必须一模一样** ——
#: 任何措辞上的差别都会泄露"这个 id 到底存不存在"。
_NOT_FOUND = "这条待办不存在。"


@router.post(
    "",
    response_model=TodoItem,
    status_code=status.HTTP_201_CREATED,
    summary="新建待办",
)
def create_todo(
    payload: TodoCreate,
    db: Session = Depends(get_db),
    learner_id: str = Depends(require_learner_id),
) -> TodoItem:
    """新建一条待办。标题的清洗与"非空"由 schema 层保证（空白标题进不来）。"""
    todo = todo_service.create_todo(
        db,
        learner_id=learner_id,
        title=payload.title,
        due_date=payload.due_date,
        timer_mode=payload.timer_mode,
        timer_minutes=payload.timer_minutes,
    )
    return TodoItem.model_validate(todo)


@router.get("", response_model=TodoListResponse, summary="我的待办")
def list_todos(
    completed: bool | None = Query(None, description="只看已完成 / 只看未完成"),
    limit: int = Query(todo_service.DEFAULT_LIMIT, ge=1, le=todo_service.MAX_LIMIT),
    offset: int = Query(0, ge=0),
    db: Session = Depends(get_db),
    learner_id: str = Depends(require_learner_id),
) -> TodoListResponse:
    """列出**自己的**待办，未完成在前、新的在上。"""
    items, total = todo_service.list_todos(
        db, learner_id=learner_id, completed=completed, limit=limit, offset=offset
    )
    return TodoListResponse(items=[TodoItem.model_validate(t) for t in items], total=total)


@router.patch("/{todo_id}", response_model=TodoItem, summary="改一条待办")
def update_todo(
    todo_id: int,
    payload: TodoUpdate,
    db: Session = Depends(get_db),
    learner_id: str = Depends(require_learner_id),
) -> TodoItem:
    """改标题 / 完成状态 / 截止日。

    ⚠️ **只改请求体里真正出现过的字段**（`exclude_unset=True`）——
    这样 `{"title": "x"}` 不会动截止日，而 `{"due_date": null}` 才是主动清空。
    两者在模型上长得一样，只有"有没有被提供"能区分。
    """
    changes = payload.model_dump(exclude_unset=True)
    if not changes:
        # 什么都没提供 ── 不是错误，只是没有可改的。返回当前状态，避免多余写入。
        return TodoItem.model_validate(
            todo_service.get_owned(db, todo_id, learner_id)
        )
    try:
        todo = todo_service.update_todo(
            db, todo_id=todo_id, learner_id=learner_id, changes=changes
        )
    except todo_service.TodoNotFound:
        raise HTTPException(status_code=404, detail=_NOT_FOUND) from None
    return TodoItem.model_validate(todo)


@router.delete("/{todo_id}", summary="删一条待办")
def delete_todo(
    todo_id: int,
    db: Session = Depends(get_db),
    learner_id: str = Depends(require_learner_id),
) -> dict:
    try:
        todo_service.delete_todo(db, todo_id=todo_id, learner_id=learner_id)
    except todo_service.TodoNotFound:
        raise HTTPException(status_code=404, detail=_NOT_FOUND) from None
    return {"deleted": True, "todo_id": todo_id}
